### PR TITLE
Expand PC commands to accept size argument

### PR DESCRIPTION
**Pull request by haruInDisguise**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**
This commit expands the behavior of the 'pc' commands, allowing them to accept an optional 'size' argument. This commit is work in progress and feedback would be very much appreciated!
In short, this feature allows the user to generate variable length arrays of data in their desired language.
The behavior of the 'pcH' etc. commands is not altered.

This addresses #3732

...

**Test plan**

Use any of the 'pc' commands with a size argument.
...

**Closing issues**

...

### CODE REVIEW

File: librz/main/rz-gg.c [hunk 1]
```diff
@@ -536,12 +536,12 @@ RZ_API int rz_main_rz_gg(int argc, const char **argv) {
 			const ut8 *tmp = rz_buf_data(b, &tmpsz);
 			switch (*format) {
 			case 'c':
-				code = rz_lang_byte_array(tmp, tmpsz, RZ_LANG_BYTE_ARRAY_C_CPP_BYTES);
+                code = rz_lang_byte_array(tmp, tmpsz, RZ_LANG_BYTE_ARRAY_BASH);
```
Comment (XVilka) [inline librz/main/rz-gg.c]: clang-format

File: librz/util/lang_byte_array.c [hunk 2]
```diff
@@ -228,6 +228,15 @@ RZ_API RZ_OWN char *rz_lang_byte_array(RZ_NONNULL const ut8 *buffer, size_t size
 	RzStrBuf sb;
 	rz_strbuf_init(&sb);
 
+	if(size == 0) {
+		RZ_LOG_ERROR("Length may not be 0\n");
+		return rz_strbuf_drain_nofree(&sb);
+	}
+
+	if(size < 0) {
```
Comment (XVilka) [inline librz/util/lang_byte_array.c]: No need, just show error on negative sizes too

File: librz/include/rz_util/rz_lang_byte_array.h [hunk 3]
```diff
@@ -34,6 +34,10 @@ typedef enum rz_lang_byte_array_type_t {
 	RZ_LANG_BYTE_ARRAY_YARA,
 } RzLangByteArrayType;
 
+typedef enum rz_lang_n_byte_array_type_t {
```
Comment (XVilka) [inline librz/include/rz_util/rz_lang_byte_array.h]: Unnecessary

File: librz/core/cmd_descs/cmd_print.yaml [hunk 4]
```diff
@@ -99,59 +102,101 @@ commands:
...
+        summary: Generate a Yara match pattern. Provide an argument to only print <N> bytes
         cname: cmd_print_byte_array_yara
-        args: []
+        args:
+          - name: N
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
       - name: pc*
-        summary: Generate a rizin commands for writing the byte array.
+        summary: Generate a rizin commands for writing the byte array. Provide an argument to only print <N> bytes.
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_print.yaml]: Shorten it into "Generate a rizin ... the byte array of <N> bytes"

Comment (XVilka) [reply librz/core/cmd_descs/cmd_print.yaml]: Similarly other summaries too

File: librz/core/cmd/cmd_print.c [hunk 5]
```diff
@@ -3642,7 +3648,12 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_yara_handler, RZ_LAN
 
 RZ_IPI RzCmdStatus rz_cmd_print_byte_array_with_inst_handler(RzCore *core, int argc, const char **argv) {
 	rz_core_block_read(core);
-	char *code = rz_core_print_bytes_with_inst(core, core->block, core->offset, (int)core->blocksize);
+    const int size = argc > 1 ? rz_num_math(core->num, argv[1]) : core->blocksize;
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c]: Run `sys/clang-format.py`

File: librz/core/cmd/cmd_print.c [hunk 6]
```diff
@@ -3642,7 +3648,12 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_yara_handler, RZ_LAN
...
+    if(size == 0) {
```
Comment (XVilka) [inline librz/core/cmd/cmd_print.c]: Check for negatives here too?

**Review by XVilka — CHANGES_REQUESTED** [review]
This idea is better, thanks

**Comment by haruInDisguise** [comment]
Thanks a lot for the review! I've implemented your suggestions and merged the dev branch.

File: librz/analysis/arch/arm/arm_il32.c [hunk 7]
```diff
@@ -1925,7 +1925,7 @@ static RzILOpEffect *sadd16(cs_insn *insn, bool is_thumb) {
 	bool is_signed = insn->id == ARM_INS_SADD16 || insn->id == ARM_INS_SHADD16 || insn->id == ARM_INS_SASX ||
 		insn->id == ARM_INS_SSAX || insn->id == ARM_INS_SHASX || insn->id == ARM_INS_SHSAX ||
 		insn->id == ARM_INS_SSUB16 || insn->id == ARM_INS_SHSUB16;
-	RzILOpBitVector *(*cast)(ut32 length, RzILOpBitVector *val) = is_signed ? rz_il_op_new_signed : rz_il_op_new_unsigned;
+	RzILOpBitVector *(*cast)(ut32 length, RzILOpBitVector * val) = is_signed ? rz_il_op_new_signed : rz_il_op_new_unsigned;
```
Comment (XVilka) [inline librz/analysis/arch/arm/arm_il32.c]: Undo changes in unrelated to your change files. What clang-format version did you use?

**Review by XVilka — CHANGES_REQUESTED** [review]
Please add/modify tests in `test/db/cmd` - you can grep for `pc` commands there to find where they are checked

File: librz/analysis/arch/arm/arm_il32.c (hunk 7 above)
Comment (haruInDisguise) [reply librz/analysis/arch/arm/arm_il32.c]: Done. My version is 15.0.7

File: librz/core/cmd_descs/cmd_print.yaml [hunk 8]
```diff
@@ -99,59 +102,101 @@ commands:
...
         summary: Generate a Yara match pattern.
         cname: cmd_print_byte_array_yara
-        args: []
+        args:
+          - name: N
+            type: RZ_CMD_ARG_TYPE_RZNUM
+            optional: true
       - name: pc*
-        summary: Generate a rizin commands for writing the byte array.
+        summary: Generate a rizin commands for writing the byte array..
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_print.yaml]: Unnecessary dot

File: librz/main/rz-gg.c [hunk 9]
```diff
@@ -536,7 +536,7 @@ RZ_API int rz_main_rz_gg(int argc, const char **argv) {
 			const ut8 *tmp = rz_buf_data(b, &tmpsz);
 			switch (*format) {
 			case 'c':
-				code = rz_lang_byte_array(tmp, tmpsz, RZ_LANG_BYTE_ARRAY_C_CPP_BYTES);
+				code = rz_lang_byte_array(tmp, tmpsz, RZ_LANG_BYTE_ARRAY_BASH);
```
Comment (XVilka) [inline librz/main/rz-gg.c]: Why? This is likely wrong

File: librz/core/cmd_descs/cmd_descs.c [hunk 10]
```diff
@@ -12596,106 +12625,197 @@ static const RzCmdDescHelp cmd_print_byte_array_asm_help = {
...
+		.type = RZ_CMD_ARG_TYPE_RZNUM,
+		.flags = RZ_CMD_ARG_FLAG_LAST,
+		.optional = true,
+
+	},
 	{ 0 },
 };
 static const RzCmdDescHelp cmd_print_byte_array_java_help = {
-	.summary = "Generate a Java byte array.",
+	.summary = "Generate a Java byte array. Provide an argument to only print <N> bytes",
```
Comment (XVilka) [inline librz/core/cmd_descs/cmd_descs.c]: You forgot to regenerate C files

**Review by XVilka — CHANGES_REQUESTED** [review]
Modify/add tests please

**Comment by haruInDisguise** [comment]
> Please add/modify tests in `test/db/cmd` - you can grep for `pc` commands there to find where they are checked

I've added tests that use the new size argument instead of specifying a block size. These changes were pushed prematurely and some of the tests fail. I will investigate this. 
Edit: It turns out I forgot to specify the newly build rizin binary on the command line. All the tests passed on my system.

**Comment by wargio** [comment]
why not just use the `@!` modifier?
```
[0x00005f70]> pc @!155
#define ARRAY_SIZE 155
const uint8_t array[ARRAY_SIZE] = {
  0xf3, 0x0f, 0x1e, 0xfa, 0x31, 0xed, 0x49, 0x89, 0xd1, 0x5e, 0x48, 0x89, 0xe2, 0x48, 0x83, 0xe4,
  0xf0, 0x50, 0x54, 0x45, 0x31, 0xc0, 0x31, 0xc9, 0x48, 0x8d, 0x3d, 0xd1, 0xe0, 0xff, 0xff, 0xff,
  0x15, 0x13, 0xbd, 0x01, 0x00, 0xf4, 0x66, 0x2e, 0x0f, 0x1f, 0x84, 0x00, 0x00, 0x00, 0x00, 0x00,
  0x48, 0x8d, 0x3d, 0xd1, 0xc2, 0x01, 0x00, 0x48, 0x8d, 0x05, 0xca, 0xc2, 0x01, 0x00, 0x48, 0x39,
  0xf8, 0x74, 0x15, 0x48, 0x8b, 0x05, 0x0e, 0xbd, 0x01, 0x00, 0x48, 0x85, 0xc0, 0x74, 0x09, 0xff,
  0xe0, 0x0f, 0x1f, 0x80, 0x00, 0x00, 0x00, 0x00, 0xc3, 0x0f, 0x1f, 0x80, 0x00, 0x00, 0x00, 0x00,
  0x48, 0x8d, 0x3d, 0xa1, 0xc2, 0x01, 0x00, 0x48, 0x8d, 0x35, 0x9a, 0xc2, 0x01, 0x00, 0x48, 0x29,
  0xfe, 0x48, 0x89, 0xf0, 0x48, 0xc1, 0xee, 0x3f, 0x48, 0xc1, 0xf8, 0x03, 0x48, 0x01, 0xc6, 0x48,
  0xd1, 0xfe, 0x74, 0x14, 0x48, 0x8b, 0x05, 0xad, 0xbf, 0x01, 0x00, 0x48, 0x85, 0xc0, 0x74, 0x08,
  0xff, 0xe0, 0x66, 0x0f, 0x1f, 0x44, 0x00, 0x00, 0xc3, 0x0f, 0x1f
};
```

**Review by wargio — CHANGES_REQUESTED** [review]
@ret2libc

**Comment by XVilka** [comment]
> why not just use the `@!` modifier?

It's clunky, unintuitive, and might be gone during the refactoring of the `core->block`

**Comment by wargio** [comment]
> It's clunky, unintuitive, and might be gone during the refactoring of the `core->block`

I agree, but at the same time its kinda redundant.

**Comment by haruInDisguise** [comment]
I don't have a solution, but my perspective might be helpful. I was completely unaware of the @! Modifier, before I began to work on this PR, and I've been using rizin/cutter for quite a while now.
To me, it seems that a bit of redundancy in a program that is as complex as rizin, could be well worth it, if that aids in making the commands a bit more intuitive and understandable.

**Comment by ret2libc** [comment]
Yeah I think we chose at the beginning to use `@!` for consistency, but at the same time it's a bit counterintuitive. I'm ok with introducing `len`, but we would also need to kill `core->block` at some point.

**Comment by wargio** [comment]
ok, then let's try to remove block size in the next release to make this consistent.

**Review by wargio — APPROVED** [review]

**Comment by XVilka** [comment]
This is the tracking issue for `core->block` refactoring, just for the record: https://github.com/rizinorg/rizin/issues/2695

**Review by XVilka — APPROVED** [review]

File: librz/core/cmd_descs/cmd_print.yaml [hunk 11]
```diff
@@ -75,15 +75,18 @@ commands:
...
-          - text: "pcj @! 10"
+          - text: "pcj 10"
             comment: "Generate a JSON bytes array of size 10"
     subcommands:
       - name: pc
         summary: Generate a C/C++ byte array.
         cname: cmd_print_byte_array_c_cpp_bytes
-        args: []
+        args:
+          - name: N
```
Comment (ret2libc) [inline librz/core/cmd_descs/cmd_print.yaml]: can we use `len` for this argument? I think this is what it's usually called in many `p?` and `px?` commands.

File: librz/main/rz-gg.c (hunk 9 above)
Comment (ret2libc) [reply librz/main/rz-gg.c]: This?

File: librz/core/cmd/cmd_print.c [hunk 12]
```diff
@@ -3642,7 +3652,12 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_yara_handler, RZ_LAN
 
 RZ_IPI RzCmdStatus rz_cmd_print_byte_array_with_inst_handler(RzCore *core, int argc, const char **argv) {
 	rz_core_block_read(core);
-	char *code = rz_core_print_bytes_with_inst(core, core->block, core->offset, (int)core->blocksize);
+	const int size = argc > 1 ? rz_num_math(core->num, argv[1]) : core->blocksize;
+	if (size <= 0) {
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c:3660]:
```
		if (size > core->blocksize_max) { \
			RZ_LOG_ERROR("Size exceeds max size (%u)\n", core->blocksize_max); \
			return RZ_CMD_STATUS_ERROR; \
		} \
```
This here as well?

**Review by ret2libc — CHANGES_REQUESTED** [review]

File: librz/main/rz-gg.c (hunk 9 above)
Comment (haruInDisguise) [reply librz/main/rz-gg.c]: Sorry, I somehow totally missed this one.

File: librz/core/cmd/cmd_print.c [hunk 13]
```diff
@@ -3642,7 +3652,16 @@ CMD_PRINT_BYTE_ARRAY_HANDLER_NORMAL(rz_cmd_print_byte_array_yara_handler, RZ_LAN
 
 RZ_IPI RzCmdStatus rz_cmd_print_byte_array_with_inst_handler(RzCore *core, int argc, const char **argv) {
 	rz_core_block_read(core);
-	char *code = rz_core_print_bytes_with_inst(core, core->block, core->offset, (int)core->blocksize);
+	const int size = argc > 1 ? rz_num_math(core->num, argv[1]) : core->blocksize;
+    if(size > core->blocksize_max) {
+			RZ_LOG_ERROR("Size exceeds max size (%u)\n", core->blocksize_max);
+			return RZ_CMD_STATUS_ERROR;
+		}
+	if (size <= 0) {
```
Comment (ret2libc) [inline librz/core/cmd/cmd_print.c]: fix style please

File: librz/core/cmd_descs/cmd_print.yaml (hunk 11 above)
Comment (ret2libc) [reply librz/core/cmd_descs/cmd_print.yaml]: why resolved? i still see `N`.

**Review by ret2libc — CHANGES_REQUESTED** [review]

**Review by XVilka — APPROVED** [review]

**Review by ret2libc — APPROVED** [review]
